src/gui/torrentcategorydialog.cpp

```diff
@@ -0,0 +1,124 @@
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "torrentcategorydialog.h"
+
+#include <QMap>
+#include <QMessageBox>
+
+#include "base/bittorrent/session.h"
+#include "ui_torrentcategorydialog.h"
+
+TorrentCategoryDialog::TorrentCategoryDialog(QWidget *parent)
+    : QDialog {parent}
+    , m_ui {new Ui::TorrentCategoryDialog}
+{
+    m_ui->setupUi(this);
+    m_ui->comboSavePath->setMode(FileSystemPathEdit::Mode::DirectorySave);
+    m_ui->comboSavePath->setDialogCaption(tr("Choose save path"));
+}
+
+TorrentCategoryDialog::~TorrentCategoryDialog()
+{
+    delete m_ui;
+}
+
+QString TorrentCategoryDialog::createCategory(QWidget *parent, const QString &parentCategoryName)
+{
+    using BitTorrent::Session;
+
+    QString newCategoryName {parentCategoryName.trimmed()};
+    if (!newCategoryName.isEmpty())
+        newCategoryName += QLatin1Char('/');
+    newCategoryName += tr("New Category");
+
+    TorrentCategoryDialog dialog(parent);
+    dialog.setCategoryName(newCategoryName);
+    while (dialog.exec() == TorrentCategoryDialog::Accepted) {
+        newCategoryName = dialog.categoryName();
+
+        if (!BitTorrent::Session::isValidCategoryName(newCategoryName)) {
+            QMessageBox::warning(
+                        parent, tr("Invalid category name")
+                        , tr("Category name must not contain '\\'.\n"
+                             "Category name must not start/end with '/'.\n"
+                             "Category name must not contain '//' sequence."));
+        }
+        else if (BitTorrent::Session::instance()->categories().contains(newCategoryName)) {
+            QMessageBox::warning(
+                        parent, tr("Category exists")
```

Comment:
"Name exists"? Not sure, but sounds clearer to me.

Reply:
"Name exists" is too abstract?
We have the "category with given name already exists" event, isn't it?

Reply:
Exactly, but the title says that we have created a category already, and then found that it duplicates already existing one. While in fact no creation happened. I agree that "name exists" is too general, but "Category exists" is simply wrong. Maybe "Unacceptable category name" or something like that?

Reply:
Maybe "Create Category failed" then?

Reply:
This looks better. Since it is the title of the dialog, maybe "Category creation error" and QMessageBox::error()?

Reply:
Ok with suggested name. But I don't want to use "error" dialog. It's not anything terrible enough to frighten the user. He can just change the name.
Or do you have serious reasons to insist on this?

Reply:
After a second thought I insist on error, because warning means that the requested action was completed, but with an undesired side-effect, or not completely. In other words, this is a type of message that can be skipped with a mood "I don't care".  But here the action failed completely. Also, we need to add an explanation how to correct the error, e.g. "Please choose a different name".

Reply:
Ok.
Then the type of message box above should be changed as well?

Reply:
Oh, sorry, I meant the QMessageBox function in the first place when was writing about the error. So yes, I strongly suggest to use `QMessageBox::critical()`.

Reply:
No. I asked about the first QMessageBox::warning with "Invalid category name" title. Should it also be changet to `critical`?

Reply:
I think yes, why would they differ?

Reply:
Ok. Changed both.